File: examples/ch07-jdbc-plus/src/main/java/flashcard/plus/domain/Card.java
package flashcard.plus.domain;

import java.time.LocalDateTime;

import org.springframework.data.annotation.Id;
import org.springframework.data.relational.core.mapping.Table;

// tag::class[]
@Table("CARD")
public record Card(@Id Long id, Long deckId, String text, String meaning,
                   LocalDateTime createdAt) {

    public static Card create(Long deckId, String text, String meaning, LocalDateTime now) {
        return new Card(null, deckId, text, meaning, now);
    }

    public Card edit(String newText, String newMeaning) {
        return new Card(id, deckId, newText, newMeaning, createdAt);
    }
}
// end::class[]

File: examples/ch08-kotlin/src/main/kotlin/flashcard/kotlin/Card.kt
package flashcard.kotlin

// tag::class[]
/**
 * Java record에 해당하는 Kotlin data class.
 * copy()로 일부 값만 바꾼 복사본을 만들 수 있고,
 * 기본값 덕분에 새 카드를 만들 때 id를 생략할 수 있다.
 */
data class Card(
    val id: Long? = null,
    val deckId: Long,
    val text: String,
    val meaning: String,
)
// end::class[]
